### PR TITLE
feat: add interceptor for bufbuild/protovalidate

### PR DESCRIPTION
**Pull request by gvencadze**
Interceptor for validating incoming requests using  [bufbuild/protovalidate](https://github.com/bufbuild/protovalidate) with Google's Common Expression Language ([CEL](https://github.com/google/cel-spec)) support (+ legacy protoc-gen-validate)

## Changes

1. Add interceptor that validates incoming requests using [bufbuild/protovalidate-go](https://github.com/bufbuild/protovalidate-go) under the hood
2. For unit-testing purposes I've added new proto contract that uses annotations from [protovalidate](https://github.com/bufbuild/protovalidate), because it doesn't generate `Validate()` interface, so I can't reuse already existing contract and codegen
3. Add [protovalidate](https://github.com/bufbuild/protovalidate) to `buf.yaml` as depedency, I've tried to use it as local, but without any success. (If usage of protoc for codegen is important, I'll rework this moment)

## Verification

I've created sample project, with some proto APIs and check both modes of [protovalidate](https://github.com/bufbuild/protovalidate) - legacy (like in protoc-gen-validate) and new powered by Google's Common Expression Language ([CEL](https://github.com/google/cel-spec))

I've added examples only for new mode

### CODE REVIEW

File: interceptors/protovalidate/doc.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+		}
+
+		message GetInstanceResponse {}
+
+
+		message LegacyRequest {
+		  string email = 1 [(validate.rules).string.email = true]; // https://github.com/bufbuild/protoc-gen-validate
+		}
+
+		message LegacyResponse {}
```
Comment (johanbrandhorst) [inline interceptors/protovalidate/doc.go:44]: Code blocks in Go comments are just intended by one space: https://go.dev/doc/comment#code.
```suggestion

 syntax = "proto3";
 package cloud.instance.v1;

 import "buf/validate/validate.proto";
 import "validate/validate.proto";

 service InstanceService {
    // GetInstance is an example of request that uses a new constraints
    rpc GetInstance(GetInstanceRequest) returns (GetInstanceResponse) {}

   // Legacy is an example of request that uses protoc-gen-validate constraints
   // their support enabled in protovalidate library constructor
   rpc Legacy(LegacyRequest) returns (LegacyResponse) {}
 }

 message GetInstanceRequest {
   string instance_id = 1 [(buf.validate.field).string.uuid = true];
 }

 message GetInstanceResponse {}

 message LegacyRequest {
   string email = 1 [(validate.rules).string.email = true]; // https://github.com/bufbuild/protoc-gen-validate
 }

 message LegacyResponse {}
```

File: interceptors/protovalidate/example_stream_test.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) The go-grpc-middleware Authors.
+// Licensed under the Apache License 2.0.
+
+package protovalidate_test
+
+import (
+	"github.com/bufbuild/protovalidate-go"
+	protovalidate_middleware "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/protovalidate"
```
Comment (johanbrandhorst) [inline interceptors/protovalidate/example_stream_test.go:8]: I don't know that we need to alias this, do we?

File: interceptors/protovalidate/example_unary_test.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) The go-grpc-middleware Authors.
+// Licensed under the Apache License 2.0.
+
+package protovalidate_test
+
+import (
+	"context"
+	"github.com/bufbuild/protovalidate-go"
+	protovalidate_middleware "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/protovalidate"
```
Comment (johanbrandhorst) [inline interceptors/protovalidate/example_unary_test.go:9]: Ditto remove the alias

File: interceptors/protovalidate/example_stream_test.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	testvalidatev1 "github.com/grpc-ecosystem/go-grpc-middleware/v2/testing/testvalidate/v1"
+	"google.golang.org/grpc"
+	"net"
```
Comment (johanbrandhorst) [inline interceptors/protovalidate/example_stream_test.go:11]: Please format these files with `goimports` to separate the import groups.

File: interceptors/protovalidate/protovalidate_test.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	sWithNoArgs := &ValidatorTestSuite{
+		InterceptorTestSuite: &testvalidate.InterceptorTestSuite{
+			ServerOpts: []grpc.ServerOption{
+				grpc.StreamInterceptor(protovalidate_middleware.StreamServerInterceptor(validator)),
+				grpc.UnaryInterceptor(protovalidate_middleware.UnaryServerInterceptor(validator)),
+			},
+		},
+	}
+	suite.Run(t, sWithNoArgs)
+}
```
Comment (johanbrandhorst) [inline interceptors/protovalidate/protovalidate_test.go:66]: Please don't use testify suites. Could you rewrite this using Go subtests instead? E.g.

```go
t.Run("TestInvalidEmail_Stream", func(t *testing.T) {
...
}
```

File: testing/testvalidate/interceptor_suite.go [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+		Type:  "CERTIFICATE",
+		Bytes: derBytes,
+	})
+	keyOut := pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(priv),
+	})
+
+	return certOut, keyOut, nil
+}
```
Comment (johanbrandhorst) [inline testing/testvalidate/interceptor_suite.go:207]: I don't think we need all of this stuff, just do the setup necessary in the test directly. Lets not bother with the TLS part of the test either. All we need is to set up a gRPC server with the interceptors registered (start it in a goroutine and use `t.Cleanup` to shut it down). The client can run in the main testing goroutine.

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
Thanks for this, this is great!

File: Makefile [hunk 7]
```diff
@@ -134,3 +134,12 @@ proto: $(BUF) $(PROTOC_GEN_GO) $(PROTOC_GEN_GO_GRPC) $(PROTO_TEST_DIR)/test.prot
...
 	    $(PROTO_TEST_DIR)/*.proto
+
+.PHONY: buf.gen
+buf.gen:
+	@docker run \
+       --volume ".:/workspace" \
+       --workdir /workspace \
+       bufbuild/buf:1.26.1 generate \
+       --template /workspace/testing/testvalidate/testvalidate.buf.gen.yaml \
+       --path /workspace/testing/testvalidate/v1
```
Comment (johanbrandhorst) [inline Makefile:145]: Just use `go run`
```suggestion
	@go run github.com/bufbuild/buf/cmd/buf@v1.26.1 generate \
       --template /workspace/testing/testvalidate/testvalidate.buf.gen.yaml \
       --path /workspace/testing/testvalidate/v1
```

File: interceptors/protovalidate/example_stream_test.go (hunk 2 above)
Comment (gvencadze) [reply interceptors/protovalidate/example_stream_test.go:8]: Idk why, but `protovalidate-go` package defined as `protovalidate`, so I've set import alias.

I'd rather would suggest to rename my package to something like `bufvalidate`, what do you think about these?

File: Makefile (hunk 7 above)
Comment (gvencadze) [reply Makefile]: Done

File: testing/testvalidate/interceptor_suite.go (hunk 6 above)
Comment (gvencadze) [reply testing/testvalidate/interceptor_suite.go]: Done

File: interceptors/protovalidate/protovalidate_test.go (hunk 5 above)
Comment (gvencadze) [reply interceptors/protovalidate/protovalidate_test.go]: done

File: interceptors/protovalidate/doc.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+/*
+Package protovalidate is a request validator that uses https://github.com/bufbuild/protovalidate-go under the hood.
+
+In case of a validation failure, an `InvalidArgument` gRPC status is returned, along with a
+description of the validation failure.
+
+It supports two ways of work:
+
+1. use new annotations that will be catched and processed by `protovalidate-go` package.
```
[the remark this answers is not in the document]
Comment (gvencadze) [reply interceptors/protovalidate/doc.go:12]: done

File: interceptors/protovalidate/doc.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+2. use legacy mode, annotations will be same as for `protoc-gen-validate` and `Validate()` method will be generated.
```
[the remark this answers is not in the document]
Comment (gvencadze) [reply interceptors/protovalidate/doc.go:14]: done

File: interceptors/protovalidate/doc.go (hunk 1 above)
Comment (gvencadze) [reply interceptors/protovalidate/doc.go]: I've removed extra spaces, so probably it should be okay now. In godoc looks okay

File: interceptors/protovalidate/example_stream_test.go (hunk 2 above)
Comment (johanbrandhorst) [reply interceptors/protovalidate/example_stream_test.go:10]: I'm confused, can't we just use it as `protovalidate`?

Comment (johanbrandhorst) [reply interceptors/protovalidate/example_stream_test.go:10]: Oh sorry, I see what you mean now. This is fine.

**Review by johanbrandhorst — APPROVED** [review]
Perfect, thanks a lot for this :)

File: Makefile [hunk 10]
```diff
@@ -134,3 +134,9 @@ proto: $(BUF) $(PROTOC_GEN_GO) $(PROTOC_GEN_GO_GRPC) $(PROTO_TEST_DIR)/test.prot
 		--go_out=$(PROTO_TEST_DIR)/../ \
 		--go-grpc_out=$(PROTO_TEST_DIR)/../ \
 	    $(PROTO_TEST_DIR)/*.proto
+
+.PHONY: buf.gen
+buf.gen:
+	@go run github.com/bufbuild/buf/cmd/buf@v1.26.1 generate \
+           --template ./testing/testvalidate/testvalidate.buf.gen.yaml \
+           --path ./testing/testvalidate/v1
```
Comment (johanbrandhorst) [inline Makefile:142]: On second thought, we actually already have `buf` available as `$(BUF)`. I think we probably need to update the version used though. Could you try using `$(BUF)`? If it doesn't work please upgrade the version we're using. I'm happy to help with this.

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]

File: Makefile (hunk 10 above)
Comment (gvencadze) [reply Makefile]: I've replaced go run ... with $(BUF) + upgrade buf to v1.26.1.

Probably it will fix also lint step in CI

Comment (gvencadze) [reply Makefile]: Hmm, lint step still failing, but now says that there's no license header in `*.pb.go` files. Can you please take a look and help me a little bit?

**Comment by johanbrandhorst** [comment]
The lint error seems to imply we need to rerun the generation script?

https://github.com/grpc-ecosystem/go-grpc-middleware/actions/runs/5872315640/job/15923608916?pr=614

File: Makefile (hunk 10 above)
Comment (johanbrandhorst) [reply Makefile]: Hm, yes, let me see

Comment (johanbrandhorst) [reply Makefile]: That should do it

**Review by johanbrandhorst — APPROVED** [review]

**Comment by johanbrandhorst** [comment]
Thanks for your contribution!

**Comment by akshayjshah** [comment]
@johanbrandhorst and @gvencadze - awesome to see this has already landed! Are you open to a follow-up PR that adds an unimplemented `Option` interface to this package, so that the two exported constructors have some room to evolve without breaking back-compat?

**Comment by gvencadze** [comment]
@akshayjshah hey, sounds like a great idea

**Comment by johanbrandhorst** [comment]
Sounds reasonable, this hasn't hit a released version yet.

**Pull request merged**